### ISSUE TITLE
Integrate D2 diagrams

### ISSUE THREAD

**Issue by alixander** [gohugoio/hugo#10579]
D2 is a new, modern diagramming language, in pure Go: https://d2lang.com/tour/intro/

Comparison to mermaid: https://text-to-diagram.com/?b=mermaid

I see that Hugo uses Goldmark, which a community member of D2 has written an extension for, making it an easy integration: https://github.com/FurqanSoftware/goldmark-d2. 

Happy to help in whatever way, or even make the PR, depending on what maintainers think. Thanks!

**Comment by bep** [comment]
Thanks for this, I notice that you're not using v8 anymore, which was the showstopper last time I looked at this.

**Comment by bep** [comment]
@alixander I quickly tested this with Hugo and, while the diagrams looks great, they take some time to generate. The Hans Niemann example (which I assume is a complicated one) takes around 70ms to build on my beefy MacBook, and building a Hugo site with 1000 pages, adding 1 diagram per page adds about 10 seconds to the build. I guess we add these to the file cache, but is this something on your TODO list to improve?

**Comment by jmooring** [comment]
It looks like [Kroki may support D2](https://github.com/yuzutech/kroki/issues/1323#issuecomment-1343292927) in the near future. If that happens, site authors can use `resources.GetRemote` in a codeblock or shortcode as demonstrated [here](https://github.com/yuzutech/kroki/issues/1323#issuecomment-1343292927).

Compared to integration, performance isn't great. But at least there's a path for site authors if we choose to defer.

**Comment by bep** [comment]
@jmooring both of your links point to the same location.

**Comment by jmooring** [comment]
Corrected.

**Comment by alixander** [comment]
@bep Thanks for the quick response. 

Performance past 70ms is not a priority for us. We translated that chess example into its equivalent in Mermaid (https://github.com/terrastruct/text-to-diagram-site/tree/master/src/examples/4_chess/syntax). It's a lot slower on my machine. 

<img width="713" alt="Screen Shot 2022-12-25 at 10 23 53 AM" src="https://user-images.githubusercontent.com/3120367/209478582-31a0a646-e6d5-4667-85a3-255e6feb35e2.png">

In the 1000 pages scenario, does Hugo build them serially or is there an option to generate the diagrams in parallel?

I'm not familiar with how `resources.GetRemote` works, but would this basically defer rendering to a client-side API call or is it making the API calls and embedding the images during build? We use an API for https://play.d2lang.com, and that can be opened up for public renders like Kroki, as I'm not sure how long the integration with D2 will take there.

**Comment by bep** [comment]
>In the 1000 pages scenario, does Hugo build them serially or is there an option to generate the diagrams in parallel?

We very much build in parallel, which, when you the math gets it down to 10ms per diagram (as in serial 70ms).

>is it making the API calls and embedding the images during build?

That. And as its backed by a file cache, it's feels fast for static resources (things that doesn't change very often).

I need to think a little about what to do about this issue.

**Comment by jmooring** [comment]
@alixander Is is making the API calls and embedding the images during build.

**Comment by alixander** [comment]
gotcha, thanks guys. let me know if there's anything I can do

**Comment by alixander** [comment]
hi @bep just want to revisit this after it's been a couple months. D2's been much improved, file sizes are 5% what they were back then, and we have a lot of new features that help blogs (like [responsive dark mode](https://d2lang.com/blog/dark-mode/)).

would you be open to an integration? if it helps, I'm happy to write the PR for it.

**Comment by jmooring** [comment]
[Kroki](https://kroki.io/) now supports D2 diagrams, so the code block render hook approach is also an option.

**Comment by LarsStegman** [comment]
Hello, are there any plans to create this integration soon?

**Comment by bligneri** [comment]
Also interested to help

**Comment by wolfspyre** [comment]
Is there a way forward here? Would be pretty neat, imo.
